File: showcase/packages/langgraph-python/src/agents/agent_config_agent.py
"""LangGraph agent backing the Agent Config Object demo.

Reads three forwarded properties — tone, expertise, responseLength — from the
LangGraph run's ``RunnableConfig["configurable"]["properties"]`` dict and
builds its system prompt dynamically per turn.

The CopilotKit provider's ``properties`` prop is wired through the runtime as
``forwardedProps`` on each AG-UI run. This graph reads those with defensive
defaults (unknown / missing values fall back to the defaults) and composes the
system prompt from three small rulebooks before invoking the model.
"""

from typing import Any, Literal

from langchain_core.runnables import RunnableConfig
from langchain_openai import ChatOpenAI
from langgraph.graph import END, START, MessagesState, StateGraph


_llm: ChatOpenAI | None = None


def _get_llm() -> ChatOpenAI:
    """Lazy-instantiate the LLM so importing this module (e.g. in unit tests)
    does not require ``OPENAI_API_KEY`` to be set."""
    global _llm
    if _llm is None:
        _llm = ChatOpenAI(model="gpt-4o-mini", temperature=0.4)
    return _llm

Tone = Literal["professional", "casual", "enthusiastic"]
Expertise = Literal["beginner", "intermediate", "expert"]
ResponseLength = Literal["concise", "detailed"]

DEFAULT_TONE: Tone = "professional"
DEFAULT_EXPERTISE: Expertise = "intermediate"
DEFAULT_RESPONSE_LENGTH: ResponseLength = "concise"

VALID_TONES: set[str] = {"professional", "casual", "enthusiastic"}
VALID_EXPERTISE: set[str] = {"beginner", "intermediate", "expert"}
VALID_RESPONSE_LENGTHS: set[str] = {"concise", "detailed"}


def read_properties(config: RunnableConfig | None) -> dict[str, str]:
    """Read the forwarded ``properties`` object with defensive defaults.

    Any missing or unrecognized value falls back to the corresponding
    ``DEFAULT_*`` constant. The function never raises.
    """
    configurable = (config or {}).get("configurable", {}) or {}
    properties = configurable.get("properties", {}) or {}

    tone = properties.get("tone", DEFAULT_TONE)
    expertise = properties.get("expertise", DEFAULT_EXPERTISE)
    response_length = properties.get("responseLength", DEFAULT_RESPONSE_LENGTH)

    if tone not in VALID_TONES:
        tone = DEFAULT_TONE
    if expertise not in VALID_EXPERTISE:
        expertise = DEFAULT_EXPERTISE
    if response_length not in VALID_RESPONSE_LENGTHS:
        response_length = DEFAULT_RESPONSE_LENGTH

    return {
        "tone": tone,
        "expertise": expertise,
        "response_length": response_length,
    }


def build_system_prompt(tone: str, expertise: str, response_length: str) -> str:
    """Compose the system prompt from the three axes."""
    tone_rules = {
        "professional": (
            "Use neutral, precise language. No emoji. Short sentences."
        ),
        "casual": (
            "Use friendly, conversational language. Contractions OK. "
            "Light humor welcome."
        ),
        "enthusiastic": (
            "Use upbeat, energetic language. Exclamation points OK. Emoji OK."
        ),
    }
    expertise_rules = {
        "beginner": "Assume no prior knowledge. Define jargon. Use analogies.",
        "intermediate": (
            "Assume common terms are understood; explain specialized terms."
        ),
        "expert": (
            "Assume technical fluency. Use precise terminology. Skip basics."
        ),
    }
    length_rules = {
        "concise": "Respond in 1-3 sentences.",
        "detailed": (
            "Respond in multiple paragraphs with examples where relevant."
        ),
    }
    return (
        "You are a helpful assistant.\n\n"
        f"Tone: {tone_rules[tone]}\n"
        f"Expertise level: {expertise_rules[expertise]}\n"
        f"Response length: {length_rules[response_length]}"
    )


def call_model(
    state: MessagesState, config: RunnableConfig | None = None
) -> dict[str, Any]:
    """Single graph node — read forwarded props, build prompt, invoke LLM."""
    props = read_properties(config)
    system_prompt = build_system_prompt(
        props["tone"], props["expertise"], props["response_length"]
    )
    messages = [{"role": "system", "content": system_prompt}] + state["messages"]
    response = _get_llm().invoke(messages)
    return {"messages": [response]}


graph_builder = StateGraph(MessagesState)
graph_builder.add_node("model", call_model)
graph_builder.add_edge(START, "model")
graph_builder.add_edge("model", END)
graph = graph_builder.compile()

File: showcase/integrations/ms-agent-dotnet/agent/AgentConfigAgent.cs
using System.Diagnostics.CodeAnalysis;
using System.Runtime.CompilerServices;
using System.Text.Json;
using Microsoft.Agents.AI;
using Microsoft.Extensions.AI;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;

// AgentConfigAgent — the /agent-config demo.
//
// Reads three forwarded properties — tone, expertise, responseLength — from
// the AG-UI shared-state payload (attached as `ag_ui_state` on
// ChatClientAgentRunOptions.AdditionalProperties, matching the convention
// already used by SharedStateAgent) and builds a dynamic system prompt per
// turn.
//
// The frontend <CopilotKitProvider agent="agent-config-demo" />'s
// useAgent().setState(...) call pushes the typed config into shared state;
// this agent reads it on every run and prepends a system message that adapts
// the inner ChatClientAgent's behavior. Missing / unrecognized values fall
// back to the documented defaults — the agent never throws on malformed
// config, so a misbehaving frontend can't kill the demo.
[SuppressMessage("Performance", "CA1812:Avoid uninstantiated internal classes", Justification = "Instantiated by SalesAgentFactory")]
internal sealed class AgentConfigAgent : DelegatingAIAgent
{
    private static readonly HashSet<string> ValidTones = new(StringComparer.Ordinal)
    {
        "professional",
        "casual",
        "enthusiastic",
    };

    private static readonly HashSet<string> ValidExpertise = new(StringComparer.Ordinal)
    {
        "beginner",
        "intermediate",
        "expert",
    };

    private static readonly HashSet<string> ValidResponseLengths = new(StringComparer.Ordinal)
    {
        "concise",
        "detailed",
    };

    private const string DefaultTone = "professional";
    private const string DefaultExpertise = "intermediate";
    private const string DefaultResponseLength = "concise";

    private readonly ILogger<AgentConfigAgent> _logger;

    public AgentConfigAgent(AIAgent innerAgent, ILogger<AgentConfigAgent>? logger = null)
        : base(innerAgent)
    {
        ArgumentNullException.ThrowIfNull(innerAgent);
        _logger = logger ?? NullLogger<AgentConfigAgent>.Instance;
    }

    public override Task<AgentRunResponse> RunAsync(IEnumerable<ChatMessage> messages, AgentThread? thread = null, AgentRunOptions? options = null, CancellationToken cancellationToken = default)
    {
        return RunStreamingAsync(messages, thread, options, cancellationToken).ToAgentRunResponseAsync(cancellationToken);
    }

    public override async IAsyncEnumerable<AgentRunResponseUpdate> RunStreamingAsync(
        IEnumerable<ChatMessage> messages,
        AgentThread? thread = null,
        AgentRunOptions? options = null,
        [EnumeratorCancellation] CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(messages);

        // Materialize up-front so we can both inspect it (to read the state)
        // and forward it to the inner agent without re-enumerating a
        // single-use iterator.
        var messageList = messages as IReadOnlyList<ChatMessage> ?? messages.ToList();

        var (tone, expertise, responseLength) = ReadConfig(options);
        var systemPrompt = BuildSystemPrompt(tone, expertise, responseLength);

        _logger.LogInformation(
            "AgentConfigAgent: tone={Tone}, expertise={Expertise}, responseLength={ResponseLength}",
            tone, expertise, responseLength);

        var systemMessage = new ChatMessage(ChatRole.System, systemPrompt);
        var augmentedMessages = new List<ChatMessage>(messageList.Count + 1) { systemMessage };
        augmentedMessages.AddRange(messageList);

        await foreach (var update in InnerAgent.RunStreamingAsync(augmentedMessages, thread, options, cancellationToken).ConfigureAwait(false))
        {
            yield return update;
        }
    }

    /// <summary>
    /// Reads the forwarded config triple from the AG-UI shared-state payload
    /// attached to the run options. Any missing / unrecognized value falls
    /// back to the corresponding default constant. Never throws.
    /// </summary>
    internal static (string Tone, string Expertise, string ResponseLength) ReadConfig(AgentRunOptions? options)
    {
        if (options is not ChatClientAgentRunOptions { ChatOptions.AdditionalProperties: { } properties } ||
            !properties.TryGetValue("ag_ui_state", out JsonElement state) ||
            state.ValueKind != JsonValueKind.Object)
        {
            return (DefaultTone, DefaultExpertise, DefaultResponseLength);
        }

        var tone = ReadStringProperty(state, "tone", ValidTones, DefaultTone);
        var expertise = ReadStringProperty(state, "expertise", ValidExpertise, DefaultExpertise);
        var responseLength = ReadStringProperty(state, "responseLength", ValidResponseLengths, DefaultResponseLength);

        return (tone, expertise, responseLength);
    }

    private static string ReadStringProperty(JsonElement state, string name, HashSet<string> valid, string defaultValue)
    {
        if (!state.TryGetProperty(name, out var element) || element.ValueKind != JsonValueKind.String)
        {
            return defaultValue;
        }
        var value = element.GetString();
        return value is not null && valid.Contains(value) ? value : defaultValue;
    }

    internal static string BuildSystemPrompt(string tone, string expertise, string responseLength)
    {
        var toneRule = tone switch
        {
            "casual" => "Use friendly, conversational language. Contractions OK. Light humor welcome.",
            "enthusiastic" => "Use upbeat, energetic language. Exclamation points OK. Emoji OK.",
            _ => "Use neutral, precise language. No emoji. Short sentences.",
        };
        var expertiseRule = expertise switch
        {
            "beginner" => "Assume no prior knowledge. Define jargon. Use analogies.",
            "expert" => "Assume technical fluency. Use precise terminology. Skip basics.",
            _ => "Assume common terms are understood; explain specialized terms.",
        };
        var lengthRule = responseLength switch
        {
            "detailed" => "Respond in multiple paragraphs with examples where relevant.",
            _ => "Respond in 1-3 sentences.",
        };

        return "You are a helpful assistant.\n\n" +
            $"Tone: {toneRule}\n" +
            $"Expertise level: {expertiseRule}\n" +
            $"Response length: {lengthRule}";
    }
}
